arelle/plugin/validate/ESEF_2022/Util.py

```diff
@@ -0,0 +1,168 @@
+            headerType = "jpg"
+        elif data[:8] == b"\x89PNG\r\n\x1a\n":
+            headerType = "png"
+        elif data[:2] in (b"MM", b"II"):
+            headerType = "tiff"
+        elif data[:2] in (b"BM", b"BA"):
+            headerType = "bmp"
+        elif data[:4] == b"\x00\x00\x01\x00":
+            headerType = "ico"
+        elif data[:4] == b"\x00\x00\x02\x00":
+            headerType = "cur"
+        elif len(data) == 0:
+            headerType = "none"
+        else:
+            headerType = "unrecognized"
+        if (("gif" in imgType and headerType != "gif") or
+            (("jpg" in imgType or "jpeg" in imgType) and headerType != "jpg") or
+            ("png" in imgType and headerType != "png")):
+            modelXbrl.error("ESEF.2.5.1.imageDoesNotMatchItsFileExtension" if isFile
+                            else "ESEF.2.5.1.incorrectMIMETypeSpecified",
+                _("Image type %(imgType)s has wrong header type: %(headerType)s"),
+                modelObject=imgElt, imgType=imgType, headerType=headerType,
+                messageCodes=("ESEF.2.5.1.imageDoesNotMatchItsFileExtension", "ESEF.2.5.1.incorrectMIMETypeSpecified"))
+
+def resourcesFilePath(modelManager: ModelManager, fileName: str) -> str:
+    # resourcesDir can be in cache dir (production) or in validate/EFM/resources (for development)
+    _resourcesDir = os.path.join( os.path.dirname(__file__), "resources") # dev/testing location
+
+    if not os.path.isabs(_resourcesDir):
+        _resourcesDir = os.path.abspath(_resourcesDir)
+    if not os.path.exists(_resourcesDir): # production location
+        _resourcesDir = os.path.join(modelManager.cntlr.webCache.cacheDir, "resources", "validation", "ESEF")
+
+    return os.path.join(_resourcesDir, fileName)
+
+def loadAuthorityValidations(modelXbrl: ModelXbrl) -> list[Any] | dict[Any, Any]:
+    _file = openFileStream(modelXbrl.modelManager.cntlr, resourcesFilePath(modelXbrl.modelManager, "authority-validations.json"), 'rt', encoding='utf-8')  # type: ignore[no-untyped-call]
+    validations = json.load(_file) # {localName: date, ...}
+    _file.close()
+    return cast(Union[dict[Any, Any], list[Any]], validations)
+
+
+def checkForMultiLangDuplicates(modelXbrl):
+
+    _factConceptContextUnitHash = defaultdict(list)
+
+    for f in modelXbrl.factsInInstance:
+        if (
+            (f.isNil or getattr(f, "xValid", 0) >= VALID)
+            and f.context is not None
+            and f.concept is not None
+            and f.concept.type is not None
+            and f.concept.type.isWgnStringFactType
+        ):
+            _factConceptContextUnitHash[f.conceptContextUnitHash].append(f)
+
+    for hashEquivalentFacts in _factConceptContextUnitHash.values():
+        if len(hashEquivalentFacts) <= 1: # skip facts present only once
+            continue;
+        _aspectEqualFacts = defaultdict(dict)  # dict [(qname,lang)] of dict(cntx,unit) of [fact, fact]
```

Comment:
Required for type checking (and erroneous semicolon):
```suggestion
            continue
        _aspectEqualFacts: defaultdict[tuple[QName, str], dict[tuple[ModelContext, ModelUnit | None], list[ModelInlineFact]]] = defaultdict(dict)
```